src/ripple/protocol/impl/PublicKey.cpp

```diff
@@ -318,31 +276,55 @@ publicKeyType (Slice const& slice)
 }
 
 bool
-verify (PublicKey const& pk,
-    Slice const& m, Slice const& sig)
+verifyDigest (PublicKey const& publicKey,
+    uint256 const& digest,
+    Slice const& sig,
+    bool mustBeFullyCanonical)
 {
-    switch(pk.type())
-    {
-    default:
-    case KeyType::secp256k1:
-    {
-        sha512_half_hasher h;
-        h(m.data(), m.size());
-        auto const digest =
-            sha512_half_hasher::result_type(h);
-        return secp256k1_ecdsa_verify(
-            secp256k1Context(), digest.data(),
-                sig.data(), sig.size(),
-                    pk.data(), pk.size()) == 1;
-    }
-    case KeyType::ed25519:
+    if (publicKeyType(publicKey) != KeyType::secp256k1)
+        LogicError("sign: secp256k1 required for digest signing");
+
+    auto const canonicality = ecdsaCanonicality(sig);
+    if (! canonicality)
+        return false;
+    if (mustBeFullyCanonical && canonicality !=
+            ECDSACanonicality::fullyCanonical)
+        return false;
+    return secp256k1_ecdsa_verify(
+        secp256k1Context(), secpp(digest.data()),
+            secpp(sig.data()), sig.size(),
+                secpp(publicKey.data()), publicKey.size()) == 1;
+}
+
+bool
+verify (PublicKey const& publicKey,
+    Slice const& m,
+    Slice const& sig,
+    bool mustBeFullyCanonical)
+{
+    if (auto const type = publicKeyType(publicKey))
     {
-        if (sig.size() != 64)
-            return false;
-        return ed25519_sign_open(m.data(),
-            m.size(), pk.data(), sig.data()) == 0;
-    }
+        if (*type == KeyType::secp256k1)
+        {
+            auto const canonicality = ecdsaCanonicality(sig);
+            if (! canonicality)
+                return false;
+            if (mustBeFullyCanonical && canonicality !=
+                    ECDSACanonicality::fullyCanonical)
+                return false;
+            return verifyDigest (publicKey,
+                sha512Half(m), sig, mustBeFullyCanonical);
+        }
+        else if (*type == KeyType::ed25519)
+        {
+            if (! ed25519Canonical(sig))
+                return false;
+            return ed25519_sign_open(
+                m.data(), m.size(), publicKey.data() + 1,
```

Comment:
@ximinez great eyes: this was a bug; we add a `0xED` prefix to Ed25519 public keys so that we can distinguish them from secp256k1 public keys which we need to "strip" before calling the ed25519 functions, hence the +1 here.

@seelabs: Ideally no code outside of `verify` should care about this implementation detail, and if that's the only place, then I'm ok with leaving it as is, and adding a comment (which I should have done in the first place, since this is obvioulsy a place were something both "weird" and important is happening and the code itself isn't conveying why this is happening.